src/Avalonia.Base/Input/FocusManager.cs

```diff
@@ -245,5 +279,887 @@ private static bool IsVisible(IInputElement e)
+                                if (parent == null)
+                                    break;
+                            }
+                        }
+                        else
+                        {
+                            parent = parentElement as IInputElement;
+                        }
+                    }
+
+                    newTabStop = GetNextOrPreviousTabStopInternal(parent, current, newTabStop, false, ref currentPassed, ref currentCompare);
+
+                    if (newTabStop == null && FocusHelpers.IsPotentialTabStop(parent) && FocusHelpers.IsFocusable(parent))
+                    {
+                        if (parent is InputElement iE && KeyboardNavigation.GetTabNavigation(iE) == KeyboardNavigationMode.Cycle)
+                        {
+                            newTabStop = GetLastFocusableElement(parent, null);
+                        }
+                        else
+                        {
+                            newTabStop = parent;
+                        }
+                    }
+                    else
+                    {
+                        if (newTabStop != null && FocusHelpers.CanHaveFocusableChildren(newTabStop as AvaloniaObject))
+                        {
+                            newTabStop = GetLastFocusableElement(newTabStop, null);
+                        }
+                    }
+
+                    if (newTabStop != null)
+                        break;
+
+                    if (IsValidTabStopSearchCandidate(parent))
+                    {
+                        current = parent;
+                    }
+
+                    parent = FocusHelpers.GetFocusParent(parent);
+                    currentPassed = false;
+                }
+            }
+
+            return newTabStop;
+        }
+
+        private IInputElement? GetNextOrPreviousTabStopInternal(IInputElement? parent, IInputElement? current, IInputElement? candidate, bool findNext, ref bool currentPassed, ref IInputElement? currentCompare)
+        {
+            var newTabStop = candidate;
+            IInputElement? childStop = null;
+            int compareIndexResult = 0;
+            bool compareCurrentForPreviousElement = false;
+
+            if (IsValidTabStopSearchCandidate(current))
+            {
+                currentCompare = current;
+            }
+
+            if (parent != null)
+            {
+                bool foundCurrent = false;
+                foreach (var child in FocusHelpers.GetInputElementChildren(parent as AvaloniaObject))
+                {
+                    childStop = null;
+                    compareCurrentForPreviousElement = false;
+                    if (child == current)
+                    {
+                        foundCurrent = true;
+                        currentPassed = true;
+                        continue;
+                    }
+
+                    childStop = null;
```

Comment:
Redundant assignment, `childStop` is already `null` here (it's been assigned a view lines above)